src/OrleansRuntime/Silo/Silo.cs

```diff
@@ -588,10 +588,6 @@ private void DoStart()
                 }
                 if (logger.IsVerbose) { logger.Verbose("Message gateway service started successfully."); }
 
-                scheduler.QueueTask(clientRegistrar.Start, clientRegistrar.SchedulingContext)
-                    .WaitWithThrow(initTimeout);
```

Comment:
The other code was not calling `SetGateway`. The bug happened in the case when the silo was NOT IsGatewayNode - no gateway at all. I know it was the case since https://github.com/dotnet/orleans/issues/1811#issuecomment-224415391 had the same issue.

So we start the ClientObserverRegistrar without GW.
If there is no GW, there is no need to ClientObserverRegistrar at all. But that would require a bugger change in the silo init logic. So I shortcutted it to still create the ClientObserverRegistrar object, but call start only if GW is created.

I have no idea why the GW is not running in those 2 cases, but that is irrelevant - we allow optional non-GW silos, so we need to support this setup. That is also the reason we did not see this bug in the past - we run all silos as GW usually.